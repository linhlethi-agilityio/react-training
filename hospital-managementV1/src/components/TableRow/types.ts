import { ReactNode } from 'react';

export interface ITableRowProps {
  className?: string;
  children: ReactNode;
}
