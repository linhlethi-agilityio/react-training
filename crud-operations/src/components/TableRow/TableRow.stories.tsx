import type { ComponentStory, Meta } from '@storybook/react';
import { Table, Tbody, Td } from '@chakra-ui/react';

// Components
import TableRow from '.';

const Template: ComponentStory<typeof TableRow> = (args) => (
  <Table>
    <Tbody>
      <TableRow {...args} />
    </Tbody>
  </Table>
);

const meta: Meta<typeof Table> = {
  component: Template as any,
};

export default meta;

export const Default = Template.bind({});
const props = {
  children: <Td>inches</Td>,
};
Default.args = props;
