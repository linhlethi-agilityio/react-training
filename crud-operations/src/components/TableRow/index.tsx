import { Tr } from '@chakra-ui/react';
import { ReactNode } from 'react';

interface TableRowProps {
  children: ReactNode;
  handleClickRow?: () => void;
}

const TableRow = ({ children, handleClickRow }: TableRowProps) => {
  return (
    <Tr onClick={handleClickRow} bg="white" borderTopWidth="10px" borderColor="background.table">
      {children}
    </Tr>
  );
};

export default TableRow;
