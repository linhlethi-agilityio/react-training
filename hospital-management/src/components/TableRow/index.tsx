import styled from 'styled-components';

const StyledTableRow = styled.tr`
  vertical-align: middle;
  border-bottom: 1px solid #e8e8e8;
`;

export { StyledTableRow };
