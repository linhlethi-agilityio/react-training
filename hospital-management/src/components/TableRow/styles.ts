import { createStyles } from '@mantine/core';

const useStyles = createStyles((theme) => ({
  row: {
    '&:hover': {
      backgroundColor: theme.colors.dark[0],
    },
  },
}));

export { useStyles };
