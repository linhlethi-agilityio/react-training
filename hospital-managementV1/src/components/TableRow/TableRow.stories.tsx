import { ComponentMeta, ComponentStory } from '@storybook/react';
import { StyledTableRow } from '@components/TableRow';
import { ITableColumn, TTableAccessor } from '@components/Table/types';
import { StyledAvatar } from '@components/Avatar';
import { DownloadIcon } from '@components/Icons';
import { StyledTableCell } from '@components/TableCell';
import { ReactNode } from 'react';

export default {
  title: 'TableRow',
  component: StyledTableRow,
} as ComponentMeta<typeof StyledTableRow>;

const Template: ComponentStory<typeof StyledTableRow> = (args) => <StyledTableRow {...args} />;

export const Default = Template.bind({});

interface DataType {
  id: string;
  avatar: string;
  doctor: string;
  date: string;
  dowloadFile?: string;
}

Default.decorators = [
  (Story, context) => {
    const { args } = context;

    const data: DataType = {
      id: '2',
      avatar: 'https://images.vexels.com/content/145908/preview/male-avatar-maker-2a7919.png',
      doctor: 'Dr. Theresa Webb',
      date: '13/10/2020',
    };

    const columns: ITableColumn<DataType>[] = [
      {
        id: '1',
        width: '80px',
        textAlign: 'center',
        accessor: (data: DataType) => <StyledAvatar size="xs" src={data.avatar} />,
      },
      {
        id: '2',
        width: '232px',
        header: 'Doctor',
        accessor: 'doctor',
      },
      {
        id: '3',
        width: '35px',
        accessor: (data: DataType) => (
          <a href={data.dowloadFile}>
            <DownloadIcon size={24} />
          </a>
        ),
      },
    ];

    const renderCell = (item: DataType, accessor: TTableAccessor<DataType>): ReactNode => {
      if (typeof accessor === 'string') {
        return item[accessor] as ReactNode;
      }

      if (typeof accessor === 'function') {
        return accessor(item);
      }
    };

    args.children = (
      <StyledTableRow>
        {columns.map((columnConfig) => (
          <StyledTableCell
            key={`${data.id}-${columnConfig.id}`}
            cellWidth={columnConfig.width}
            cellTextAlign={columnConfig.textAlign}
          >
            {renderCell(data, columnConfig.accessor)}
          </StyledTableCell>
        ))}
      </StyledTableRow>
    );

    return <Story {...args} />;
  },
];
