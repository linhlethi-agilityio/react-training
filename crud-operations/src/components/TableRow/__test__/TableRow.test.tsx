import { render, fireEvent } from '@testing-library/react';
import { Table, Tbody, Td } from '@chakra-ui/react';

// Components
import TableRow from '..';

describe('TableRow', () => {
  test('should match snapshot for TableRow', () => {
    const handleClickRow = jest.fn();

    const { container } = render(
      <Table>
        <Tbody>
          <TableRow handleClickRow={handleClickRow}>
            <Td>inches</Td>
          </TableRow>
        </Tbody>
      </Table>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders children and handles click event', () => {
    // Mock handleClickRow function
    const handleClickRow = jest.fn();

    // Render TableRow with mock props
    const { getByRole } = render(
      <Table>
        <Tbody>
          <TableRow handleClickRow={handleClickRow}>
            <Td>inches</Td>
          </TableRow>
        </Tbody>
      </Table>,
    );

    // Get the rendered TableRow component
    const tableRow = getByRole('row');

    // Assert that TableRow renders children correctly
    expect(tableRow).toBeInTheDocument();
    expect(tableRow).toHaveTextContent('inches');

    // Simulate a click event on the TableRow
    fireEvent.click(tableRow);

    // Assert that handleClickRow is called when TableRow is clicked
    expect(handleClickRow).toHaveBeenCalled();
  });
});
