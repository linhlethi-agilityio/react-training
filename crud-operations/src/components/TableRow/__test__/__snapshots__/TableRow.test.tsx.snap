// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow should match snapshot for TableRow 1`] = `
<div>
  <table
    class="chakra-table css-0"
  >
    <tbody
      class="css-0"
    >
      <tr
        class="css-1w3yp13"
      >
        <td
          class="css-0"
        >
          inches
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
